### PR TITLE
Podcast Player - disallow feeds containing invalid audio types as track data.

### LINKED ISSUE #15134: Podcast Player: make sure embedded URL is truly a podcast, not a common RSS.
With the Podcast Player Block trying to embed https://distributed.blog/podcast/ will cause the block to render but the tracks will be invalid and result in an error when trying to play.

This is because the tracks end up being valid `jpeg` images rather than audio.

Screen Shot 2020-03-26 at 11 53 54

The fix should be to detect (in `_inc/lib/class-jetpack-podcast-helper.php`)  whether any tracks in the track list are not valid `wp_get_audio_extensions` and if so discard them.

Then in the track list empty return an error.

Note this is tricky because we need to detect whether the track's `url` file is anything other valid audio extensions. However some URLs are actually redirects which resolve to valid extension. How can we solve this (@obenland any ideas?)?

### PR DESCRIPTION
**Pull request by getdave**
> With the Podcast Player Block trying to embed https://distributed.blog/podcast/ will cause the block to render but the tracks will be invalid and result in an error when trying to play.

The reason for the above error is that the tracks SimplePie parses from the feed end up containing image data. This is not valid audio and therefore causes the player to error.

This PR resolves this by only allow tracks from the feed which contain valid a type which is valid audio. Anything else is filtered out.

Moreover, if there are no tracks _at all_, then the Podcast is considered invalid for embedding.

Fixes https://github.com/Automattic/jetpack/issues/15134

## Changes proposed in this Pull Request:
* Check each feed track enclosure and only allow those with valid audio.
* Error if there are no valid tracks or the tracks are empty.
* Skip caching of feed if `WP_DEBUG` constant is set.
### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* Fixes existing feature.

### On Master

* Insert Podcast Block
* Paste in `https://distributed.blog/podcast/`.
* See feed render in Block.
* Click on episodes - see that they error and cannot play.
* Inspect Network requests for REST API call - see JSON response contains tracks whose URL contains image data not audio data.

### On This Branch

* Insert Podcast Block
* Paste in `https://distributed.blog/podcast/`.
* See feed render in Block but only with items that are playable without erroring.
* Click on episodes - see that they do not error and are all playable.
* Inspect Network requests for REST API call - see JSON response contains tracks whose URL contains only audio data.

* Insert another Podcast Block.
* Insert a valid RSS feed URL such as `https://feeds.megaphone.fm/darknetdiaries`.
* See that tracks data still renders correctly for these, is playable and doesn't error.

* Insert any random URL (like `http://simplepie.org/wiki/reference/simplepie_caption/get_type`).
* See that this is still rejected as an invalid URL type.

## Proposed changelog entry for your changes:

* Fix Podcast Block to only display tracks for valid audio data.
* Skip caching of Podcast Feed on Podcast Block if `WP_DEBUG` constant is set.

### CODE REVIEW

File: _inc/lib/class-jetpack-podcast-helper.php [hunk 1]
```diff
@@ -26,16 +26,21 @@ public static function get_player_data( $feed ) {
 		$player_data   = get_transient( $transient_key );
 
 		// Fetch data if we don't have any cached.
-		if ( false === $player_data ) {
+		if ( false === $player_data || ( defined( 'WP_DEBUG' ) && WP_DEBUG ) ) {
```
Comment (getdave) [inline _inc/lib/class-jetpack-podcast-helper.php]: We can pull this into a separate PR if you'd like.

**Comment by getdave** [comment]
Is fixing the CodeClimate stuff required for this to pass. Seems fairly arbitrary and a refactor would be unrelated to this PR. 

Happy to do it but just checking...

File: _inc/lib/class-jetpack-podcast-helper.php (hunk 1 above)
Comment (marekhrabe) [reply _inc/lib/class-jetpack-podcast-helper.php]: no problemo!

**Review by marekhrabe — DISMISSED** [review]
I think this is good to go. I like the utilization of `WP_Error` throughout our functions - it makes things clear and easy to follow.

I wonder if we should straight out reject https://distributed.blog/podcast/ (which if I'm not mistaken loads RSS from https://distributed.blog/feed/), because it's content is not meant to represent a podcast. You can see it formats awkwardly, has duplicate episodes with different titles etc… Far from nice.

For now, this is however perfect to get in, as it prevents broken experience where we try to load image file in `<audio>` 👍

**Comment by getdave** [comment]
Waiting Jetpack Crew.

File: _inc/lib/class-jetpack-podcast-helper.php [hunk 2]
```diff
@@ -166,8 +179,7 @@ private static function get_audio_enclosure( SimplePie_Item $episode ) {
 			}
 		}
 
-		// Default to empty SimplePie_Enclosure object.
-		return $episode->get_enclosure();
+		return new \WP_Error( 'invalid_audio', 'Podcast audio is an invalid type.' );
```
Comment (jeherve) [inline _inc/lib/class-jetpack-podcast-helper.php]: Do we want this to be translatable?

We can probably remove `\` as well, to stay consistent with the rest of the file.

**Review by jeherve — CHANGES_REQUESTED** [review]

File: _inc/lib/class-jetpack-podcast-helper.php (hunk 2 above)
Comment (getdave) [reply _inc/lib/class-jetpack-podcast-helper.php]: You are absolutely right. Sorry I missed it!

Comment (getdave) [reply _inc/lib/class-jetpack-podcast-helper.php]: Fixed.

**Review by jeherve — APPROVED** [review]
LGTM! 🚢

**Review by marekhrabe — APPROVED** [review]
🚢

File: _inc/lib/class-jetpack-podcast-helper.php [hunk 3]
```diff
@@ -166,8 +179,7 @@ private static function get_audio_enclosure( SimplePie_Item $episode ) {
 			}
 		}
 
-		// Default to empty SimplePie_Enclosure object.
-		return $episode->get_enclosure();
+		return new WP_Error( 'invalid_audio', __( 'Podcast audio is an invalid type.', 'jetpack' ) );
```
Comment (obenland) [inline _inc/lib/class-jetpack-podcast-helper.php]: Is there a reason we return `WP_Error` here rather than `null`, like `SimplePie_Item::get_enclosure()`? The error doesn't seem to see the light of day either

**Comment by marekhrabe** [comment]
r204965-wpcom

File: _inc/lib/class-jetpack-podcast-helper.php (hunk 3 above)
Comment (getdave) [reply _inc/lib/class-jetpack-podcast-helper.php]: Yes we should have returned null particularly as the error isn't exposed. 

Did you create a followup Issue for this yet?
